***
***

# **COVID-19 Vaccination Progress**

## *1) Initial Data Exploration, Cleaning and Feature Engineering*

***
***

## Table of Contents

| 1. INTRODUCTION | 2. IMPORTS AND DATA PREPARATION | 3. SUMMARY |
|----|----|----|
| *1.1 COVID-19 Vaccines Background Information* | *2.1 Importing Libraries and Data* | *3.1 Key Findings* |
| *1.2 Dataset Characteristics* | *2.2 Preliminary Data Exploration and Subsequent Cleaning* | *3.2 Key Limitations* |
| |  *2.3 Feature Engineering to Prepare for In-Depth Data Analysis* | |
| |  *2.4 Global Vaccination Dataframe* | |

***
***

## 1. INTRODUCTION

The first vaccine to be developed against a contagious disease was the smallpox vaccine. In 1796, Edward Jenner demonstrated that inoculation with the relatively mild cowpox virus conferred immunity against the deadly smallpox virus. As the discoverer of the small pox vaccine, Edward Jenner, an English surgeon, helped save the lives of many people. These days, over 200 years later, the principle concept of vaccination remains the main option for fighting viral diseases and as such is key to combating the COVID-19 pandemic.

***

### 1.1 COVID-19 Vaccines

The speed at which the first COVID-19 vaccines were developed as a response to the pandemic was extraordinary. Overall it took less than a year to successfully develop and approve several vaccines, a timeframe significantly shorter than usually observed for the development of new vaccines. It is important that development of further vaccines continues, as it does, as a very large share of the world population will need to receive a COVID-19 vaccine in order to become immune to the virus, so that we can successfully combat the pandemic. In addition, the appearance of more and more mutated versions of the SARS-CoV-2 virus calls for the development of additional vaccines to offer a wide bandwidth of starting points to also sucessfully figth these mutants.

As such, it is evident that one of the key challenges is to vaccinate people around the world.

There are currrently four different types of vaccines developed to target SARS-CoV-2, namely inactivated viruses, recombinant viral vectors, messengerRNA and protein subunit based vaccines. An overview of the different types can be found [here](https://www.who.int/news-room/feature-stories/detail/the-race-for-a-covid-19-vaccine-explained) and [here](https://covid19.trackvaccines.org/how-do-vaccines-work/#types).

#### The two key types of vaccines utilised against SARS-CoV-2 are:

mRNA VACCINES: mRNA vaccines contain mRNA with instructions for making the spike protein. This protein is expressed the surface of the SARS-CoV-2 virus and required to enter cells. Upon vaccination the spike protein is produced and will be recognised by the immune system as foreign. This results in the production of antibodies and activated T cells (white blood cells) which will attack it. This information is stored by the immune system so that if it encounters the SARS-CoV-2 virus again it can act fast to eliminate it.

RECOMBINANT VACCINES: Recombinant vaccines are made up of another virus (which cannot reproduce and cause disease) that has been modified to contain the gene for making the SARS-CoV-2 spike protein. This protein is expressed the surface of the SARS-CoV-2 virus and required to enter cells. Upon vaccination the spike protein is produced and will be recognised by the immune system as foreign. This results in the production of antibodies and activated T cells (white blood cells) which will attack it. This information is stored by the immune system so that if it encounters the SARS-CoV-2 virus again it can act fast to eliminate it.

Several institutions maintain websites on which they list COVID-19 candidate vaccines that are currently being developed. The [WHO Vaccine Tracker](https://www.who.int/publications/m/item/draft-landscape-of-covid-19-candidate-vaccines) is one of them and provides a concise overview. Additional information on COVID-19 vaccines by the WHO can be found [here](https://www.who.int/emergencies/diseases/novel-coronavirus-2019/covid-19-vaccines).

#### General overview of the current (31st March 2021) top 4 vaccines with respect to the number of countries in which they received authorisation:

| VACCINE | Approved in | type of vaccine |
|----|----|----|
| Oxford/AstraZeneca | 82 countries | RECOMBINANT (non-replicating viral vector) |
| Pfizer/BioNTech | 79 countries | messenger RNA (mRNA) (nucleoside-modified) |
| Johnson & Johnson (Janssen) | 37 countries | RECOMBINANT (non-replicating viral vector) |
| Covishield (Serum Inst. of India) | 31 countries | RECOMBINANT (non-replicating viral vector) |

#### Overview of approved vaccines for selected countries (USA, EU, China, India, Israel, Russia) as of 31st March 2021:

#### USA

+ [FDA vaccine overview](https://www.fda.gov/emergency-preparedness-and-response/coronavirus-disease-2019-covid-19/covid-19-vaccines)

| VACCINE | status | type of vaccine | administration |
|----|----|----|----|
| Pfizer/BioNTech | authorised for emergency use (11/12/2020) | RECOMBINANT (non-replicating viral vector) | 2 intramuscular injections, 4-12 weeks apart |
| Moderna | authorised for emergency use (18/12/2020) | messenger RNA (mRNA) (nucleoside-modified) | 2 intramuscular injections, 28 days apart |
| Johnson & Johnson (Janssen) | authorised for emergency use (27/02/2021) | RECOMBINANT (non-replicating viral vector) | - |


#### European Union

NOTE: The following summary reflects the approved vaccines in the EU, not Europe. In Switzerland the Oxford/AstraZeneca vaccine is currently not approved, and in the UK Johnson & Johnson is currently not approved.

+ [EMA vaccine overview](https://www.ema.europa.eu/en/human-regulatory/overview/public-health-threats/coronavirus-disease-covid-19/treatments-vaccines/vaccines-covid-19/covid-19-vaccines-key-facts)

+ [Authorised vaccine overview - European medicines agency - EMA](https://www.ema.europa.eu/en/human-regulatory/overview/public-health-threats/coronavirus-disease-covid-19/treatments-vaccines/vaccines-covid-19/covid-19-vaccines-authorised)

| VACCINE | status | type of vaccine | administration |
|----|----|----|----|
| Pfizer/BioNTech (Corminaty) | conditional marketing authorisation granted (21/12/2020) | messenger RNA (mRNA) (nucleoside-modified) | 2 intramuscular injections, 3 weeks apart |
| Johnson & Johnson (Janssen) | conditional marketing authorisation granted (11/03/2021) | RECOMBINANT (non-replicating viral vector) | - |
| Moderna | conditional marketing authorisation granted (06/01/2021) | messenger RNA (mRNA) (nucleoside-modified) | 2 intramuscular injections, 28 days apart |
| AstraZeneca (Vaxzevria) | conditional marketing authorisation granted (29/01/2021) | RECOMBINANT (non-replicating viral vector) | 2 intramuscular injections, 4-12 weeks apart |


#### China, India, Israel and Russia

| VACCINE | type of vaccine | country |
|----|----|----|
| Sinopharm Wuhan | INACTIVATED virus | China |
| Sinopharm Beijing | INACTIVATED virus | China |
| Covaxin (Bharat Biotech) | INACTIVATED virus | China, India |
| CanSino | RECOMBINANT (non-replicating viral vector) | China |
| Anhui Zhifei Longcom (RBD-Dimer) | Protein subunit | China |
| Oxford/AstraZeneca | RECOMBINANT (non-replicating viral vector) | India |
| Covishield (Serum Inst. of India) | RECOMBINANT (non-replicating viral vector) | India |
| Pfizer/BioNTech | messenger RNA (mRNA) (nucleoside-modified) | Israel |
| Moderna | messenger RNA (mRNA) (nucleoside-modified) | Israel |
| Sputnik V | RECOMBINANT (non-replicating viral vector) | Russia |
| EpiVacCorona | Protein subunit | Russia |

An overview of clinical trials and vaccine authorisations by country can be found [here](https://covid19.trackvaccines.org/trials-vaccines-by-country/).

In summary, it can be noted that out of the **total 12 vaccines** listed above to date only 10 appear in the dataset, i.e. Covaxin, EpiVacCorona, Johnson & Johnson, Moderna, Oxford/AstraZeneca, Pfizer/BioNTech, Sinopharm Beijing, Sinopharm Wuhan, Sinovac, Sputnik V. Two chinese vaccines CanSino and Anhui Zhifei Longcom are not currently covered by the dataset. Since data for vaccinations in China is included in the dataset it can be assumed that these approved vaccines are possibly currently not being supplied yet. It can also be noted that some vaccines are approved accross numerous countries, whereas others are predominantly appoved in only a limited number of countries. The above overview of vaccines is not exhaustive and there may be some approved vaccines that are not included. And, of course, the approval of new vaccines is a dynamic progress, so the overview only reflects the status as of 31st March 2021.

IMPORTANT NOTE! Different vaccines are approved for use in different countries and thus influence which vaccines and how many are available in each country. In addition the available quantity is strongly influenced by the production capacity of each vaccine.

***

### 1.2 Dataset Overview

The dataset used in this capstone project is 'COVID-19 World Vaccination Progress (version 81)', was created by Gabriel Preda and is [available on Kaggle](https://www.kaggle.com/gpreda/covid-world-vaccination-progress). It is a compilation of COVID-19 vaccination data (updated daily) provided by [OurWordinData](https://ourworldindata.org/covid-vaccinations). The 'OurWorldinData' website contains a plethora of information regarding the global vaccination progress, including visualisations of the progress as well as an overview of approved vaccines worldwide. 

**The dataset contains 8079 observations for 15 features.**

| FEATURE | description | comment |
|----|----|----|
| **country** | 147 countries | 4 duplicates, UN recognises 197 countries (193 are members) |
| **iso_code** | country code by ISO norm | -|
| **date** | data entry (13th December 2020 to 24th March 2021) | - |
| **vaccines** | name/company | - |
| **daily_vaccinations** | no. of vaccination for that date/country | some days only daily vaccinations, some days only the (cummulative) total |
| **daily_vaccinations_raw** | daily vaccinations as CHANGE COMPARED TO PREVIOUS DAY (cross-check purposes) | calculation: total_vaccinations (day) - total_vaccination (previous day) |
| **daily_vaccinations_per_million** | ratio (in ppm) between vaccination number and total population for the current date in the country | - |
| **total_vaccinations** | absolute number of immunisations (jabs) in the country | some days only daily vaccinations, some days only the (cummulative) total |
| **total_vaccinations_per_hundred** | ratio (in percent) between vaccination number and total population up to the date in the country | - |
| **people_vaccinated** | total no. of people with immunisation (partial and full) | - |
| **people_vaccinated_per_hundred** | ratio (in percent) between population immunized (partial and full) and total population | - |
| **people_fully_vaccinated** | total no. of people with full immunisation (typically 2) | - |
| **people_fully_vaccinated_per_hundred** | ratio (in percent) between population fully immunized and total population | - |
| **source_name** | national authority, international organization, local organization etc. | - |
| **source_website** | website link to source | - |

***
***

## 2. IMPORTS AND DATA PREPARATION

***

### 2.1 Importing Libraries and Data

+ Import numpy (linear algebra) and pandas (data processing).
+ Import Kaggle COVID vaccination progress dataset.

In [1]:
#Import basic libraries
import numpy as np
import pandas as pd

#Suppress pandas future warnings
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)

In [2]:
#Import local Kaggle dataset (version 81, March 2021) country_vaccinations on COVID-19 vaccination progress and set date to datetime
df = pd.read_csv('data/country_vaccinations.csv', parse_dates=['date'])

#Alternatively load updated Kaggle dataset country_vaccinations on COVID-19 vaccination progress from url
#df = pd.read_csv('https://www.kaggle.com/gpreda/covid-world-vaccination-progress/download', parse_dates=['date'])

#Sort by date
df.sort_values('date', inplace=True)

***

### 2.2 Preliminary Data Exploration and Subsequent Cleaning

In this section the basic characteristics of the data were explored, starting with the number of observations and features, as well as the types of variables, and any missing values and/or duplicates.

NOTE: Detailed attention to missing values from a time-series perspective is given in one of the following notebooks.

Get an initial feel for the data by looking at the first 3 and last 3 rows of the dataframe:

In [3]:
#Take a first glance at the data (first 3 entries)
df.head(3)

country  iso_code       date  total_vaccinations  \
5309  Northern Ireland  OWID_NIR 2020-12-13              3623.0   
7943             Wales  OWID_WLS 2020-12-13              8257.0   
2269           England  OWID_ENG 2020-12-13             55576.0   

      people_vaccinated  people_fully_vaccinated  daily_vaccinations_raw  \
5309             3623.0                      NaN                     NaN   
7943             8257.0                      NaN                     NaN   
2269            55576.0                      NaN                     NaN   

      daily_vaccinations  total_vaccinations_per_hundred  \
5309                 NaN                            0.19   
7943                 NaN                            0.26   
2269                 NaN                            0.10   

      people_vaccinated_per_hundred  people_fully_vaccinated_per_hundred  \
5309                           0.19                                  NaN   
7943                           0.26                                  NaN   
2269                           0.10                                  NaN   

      daily_vaccinations_per_million                             vaccines  \
5309                             NaN  Oxford/AstraZeneca, Pfizer/BioNTech   
7943                             NaN  Oxford/AstraZeneca, Pfizer/BioNTech   
2269                             NaN  Oxford/AstraZeneca, Pfizer/BioNTech   

                           source_name  \
5309  Government of the United Kingdom   
7943  Government of the United Kingdom   
2269  Government of the United Kingdom   

                                         source_website  
5309  https://coronavirus.data.gov.uk/details/health...  
7943  https://coronavirus.data.gov.uk/details/health...  
2269  https://coronavirus.data.gov.uk/details/health...

In [4]:
#Take a first glance at the data (last 3 entries)
df.tail(3)

country iso_code       date  total_vaccinations  people_vaccinated  \
4959  Mongolia      MNG 2021-03-24            244550.0                NaN   
4108     Kenya      KEN 2021-03-24             64100.0            64100.0   
8078  Zimbabwe      ZWE 2021-03-24             49404.0            49404.0   

      people_fully_vaccinated  daily_vaccinations_raw  daily_vaccinations  \
4959                      NaN                 21106.0             14988.0   
4108                      NaN                     NaN              6300.0   
8078                      NaN                  5269.0              1400.0   

      total_vaccinations_per_hundred  people_vaccinated_per_hundred  \
4959                            7.46                            NaN   
4108                            0.12                           0.12   
8078                            0.33                           0.33   

      people_fully_vaccinated_per_hundred  daily_vaccinations_per_million  \
4959                                  NaN                          4572.0   
4108                                  NaN                           117.0   
8078                                  NaN                            94.0   

                vaccines                     source_name  \
4959  Oxford/AstraZeneca  Ministry of Health via ikon.mn   
4108  Oxford/AstraZeneca              Ministry of Health   
8078   Sinopharm/Beijing              Ministry of Health   

                                         source_website  
4959                                   https://ikon.mn/  
4108  https://twitter.com/MOH_Kenya/status/137473931...  
8078  https://twitter.com/MoHCCZim/status/1374795915...

**KEY FINDING:**

Some of the ISO country codes contain the prefix OWID (maybe this stands for: 'Our-World-in-Data'?).

Irrespectively, this needs to be removed and is thus our first cleaning task.

In [5]:
#Remove 'OWID' prefix from country ISO codes
df['iso_code'] = df.iso_code.map(lambda x: x.lstrip('OWID_'))
df.head(1)

country iso_code       date  total_vaccinations  \
5309  Northern Ireland      NIR 2020-12-13              3623.0   

      people_vaccinated  people_fully_vaccinated  daily_vaccinations_raw  \
5309             3623.0                      NaN                     NaN   

      daily_vaccinations  total_vaccinations_per_hundred  \
5309                 NaN                            0.19   

      people_vaccinated_per_hundred  people_fully_vaccinated_per_hundred  \
5309                           0.19                                  NaN   

      daily_vaccinations_per_million                             vaccines  \
5309                             NaN  Oxford/AstraZeneca, Pfizer/BioNTech   

                           source_name  \
5309  Government of the United Kingdom   

                                         source_website  
5309  https://coronavirus.data.gov.uk/details/health...

In [6]:
#Check if there are any more ISO country codes that are not 3-letter codes
df_iso_check = df[df['iso_code'].str.len()!=3]
print(f'ISO code cleaning (1): In total we now have {df_iso_check.shape[0]} out of {df.shape[0]} rows which are not 3-letter country ISO code conform.')
print(f'ISO code cleaning (1): There are {df_iso_check.iso_code.nunique()} country ISO codes that are not made up of 3 letters: \n {df_iso_check.iso_code.unique()}')

ISO code cleaning (1): In total we now have 1037 out of 8079 rows which are not 3-letter country ISO code conform.
ISO code cleaning (1): There are 14 country ISO codes that are not made up of 3 letters: 
 ['LS' 'SR' 'TA' 'EU' 'NK' 'MN' 'SL' 'RL' 'N' 'ND' 'ZA' 'RN' 'MA' 'M']


In [7]:
#Replace obvious 2-letters iso country codes by 3-letter codes (7 out of 14)
df['iso_code'] = df['iso_code'].replace(['LS'],'LSO').replace(['SR'],'SUR').replace(['MN'],'MNG')\
                               .replace(['SL'],'SLE').replace(['ZA'],'ZAF').replace(['MA'],'MAR')\
                               .replace(['TA'],'TAA')

In [8]:
#Check how many non-3-letter iso codes are left
df_iso_check_2 = df[df['iso_code'].str.len()!=3]
print(f'ISO code cleaning (2)): In total there are now {df_iso_check_2.shape[0]} out of {df.shape[0]} rows with non-3-letter ISO code left.')
print(f'ISO code cleaning (2): There are {df_iso_check_2.iso_code.nunique()} country ISO codes that are not made up of 3 letters: \n {df_iso_check_2.iso_code.unique()}')

ISO code cleaning (2)): In total there are now 472 out of 8079 rows with non-3-letter ISO code left.
ISO code cleaning (2): There are 7 country ISO codes that are not made up of 3 letters: 
 ['EU' 'NK' 'RL' 'N' 'ND' 'RN' 'M']


In [9]:
#Quickly check which countries belong to the remaining obscure codes
print(f'Examples of the more obscure country codes:')
df_iso_check_2[['country', 'iso_code']].sample(3, random_state=42)

Examples of the more obscure country codes:


country iso_code
3564  Ireland       RL
2720  Germany       EU
2028  Denmark       NK

In [10]:
#Replace remaining codes by 3-letter country codes
df['iso_code'] = df['iso_code'].replace(['IR'],'IRL').replace(['NK'],'DNK').replace(['EU'],'DEU')\
                               .replace(['ND'],'IND').replace(['N'],'IDN').replace(['M'],'DOM')\
                               .replace(['RN'],'IRN').replace(['RL'],'IRL')

In [11]:
#Check if any non-3-letter ISO codes left
df_iso_check_3 = df[df['iso_code'].str.len()!=3]
print(f'FINAL CHECK: In total there are {df_iso_check_3.shape[0]} out of {df.shape[0]} rows with non-3-letter ISO code left.')

FINAL CHECK: In total there are 0 out of 8079 rows with non-3-letter ISO code left.


Next, let's check out what timeframe is covered by this timeseries.

In [12]:
#Check for first/last entry date
d_min = df.date.min()
d_max = df.date.max()
print(f'The dataset covers vaccinations between {d_min} and {d_max}.')

The dataset covers vaccinations between 2020-12-13 00:00:00 and 2021-03-24 00:00:00.


In [13]:
#Check time delta
t_delta = df.date.max() - df.date.min()
print(f'The dataset covers {t_delta}.')

The dataset covers 101 days 00:00:00.


Next, let's look at the characteristics of the dataframe in a little more detail:

+ Size of the dataframe
+ Check for any duplicate entries
+ Get information wrt the data types and any missing values

In [14]:
#Number of columns and rows
shape1 = df.shape
print(f'CHECK: In total we have {shape1[1]} columns and {shape1[0]} rows in our COVID vaccination dataset.')

CHECK: In total we have 15 columns and 8079 rows in our COVID vaccination dataset.


In [15]:
#Generate count statistics of duplicate entries
if len(df[df.duplicated()]) > 0:
    print("CHECK: No. of duplicated entries in dataframe: ", len(df[df.duplicated()]))
    print(df[df.duplicated(keep=False)].sort_values(by=list(df.columns)).head())
else:
    print("CHECK: No duplicated entries found in COVID vaccination progress dataset.")

CHECK: No duplicated entries found in COVID vaccination progress dataset.


In [16]:
#Information about datatypes per column
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8079 entries, 5309 to 8078
Data columns (total 15 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   country                              8079 non-null   object        
 1   iso_code                             8079 non-null   object        
 2   date                                 8079 non-null   datetime64[ns]
 3   total_vaccinations                   4984 non-null   float64       
 4   people_vaccinated                    4423 non-null   float64       
 5   people_fully_vaccinated              3021 non-null   float64       
 6   daily_vaccinations_raw               4192 non-null   float64       
 7   daily_vaccinations                   7891 non-null   float64       
 8   total_vaccinations_per_hundred       4984 non-null   float64       
 9   people_vaccinated_per_hundred        4423 non-null   float64       
 10  people_fu

**KEY FINDINGS:**

1) The dataset contains 1 date column, 5 categorical columns and 9 columns containing numerical values.

1) NOTE: 9 out of 15 columns contain null values and need to be dealt with!

    + Therefore, let's take a look at how many empties we have to deal with.
    + And, let's check if there are any empty strings or lists in the 'object' columns. 

In [17]:
#View number of empties per column
missing = df.isnull().sum(axis = 0)
print(f'The absolute number of missing values per column are: \n \n {missing}')

The absolute number of missing values per column are: 
 
 country                                   0
iso_code                                  0
date                                      0
total_vaccinations                     3095
people_vaccinated                      3656
people_fully_vaccinated                5058
daily_vaccinations_raw                 3887
daily_vaccinations                      188
total_vaccinations_per_hundred         3095
people_vaccinated_per_hundred          3656
people_fully_vaccinated_per_hundred    5058
daily_vaccinations_per_million          188
vaccines                                  0
source_name                               0
source_website                            0
dtype: int64


In [18]:
#Percent missing per columns
percent_m = df.isnull().sum() * 100 / len(df)
print(f'The percentage of missing values per column are: \n \n {percent_m}')

The percentage of missing values per column are: 
 
 country                                 0.000000
iso_code                                0.000000
date                                    0.000000
total_vaccinations                     38.309197
people_vaccinated                      45.253125
people_fully_vaccinated                62.606758
daily_vaccinations_raw                 48.112390
daily_vaccinations                      2.327021
total_vaccinations_per_hundred         38.309197
people_vaccinated_per_hundred          45.253125
people_fully_vaccinated_per_hundred    62.606758
daily_vaccinations_per_million          2.327021
vaccines                                0.000000
source_name                             0.000000
source_website                          0.000000
dtype: float64


In [19]:
#Check for empty strings/lists in object columns
object_lst = ['country', 'iso_code', 'date', 'vaccines', 'source_name', 'source_website']
lst_empties = ['', []]
check_str_lst = df[object_lst].isin(lst_empties).value_counts()
print(f'The check for empty strings/lists in the object columns returns: \n \n {check_str_lst}')

The check for empty strings/lists in the object columns returns: 
 
 country  iso_code  date   vaccines  source_name  source_website
False    False     False  False     False        False             8079
dtype: int64


**KEY FINDINGS:**

1) There are no empty lists or strings in the dataset.

1) None of the 5 categorical columns contain empties.

1) All of the 9 numerical columns contain empties. The 2 daily vaccination columns contain relatively few empties, whereas the other columns contain 40-60% empties. Whereas the empties in the daily vaccinations will most likely be relatively easy to deal with, the large number of empties in the other columns will pose a challenge for  the time series specific analysis. This will be dealt with in one of the following notebook. 

As a next steps, let's take a look at the descriptive statistics, both for the numerical, as well as the categorical columns.

In [20]:
#Basic statistics (1)
df.describe().transpose().round()

count       mean        std  min  \
total_vaccinations                   4984.0  2597915.0  9383280.0  0.0   
people_vaccinated                    4423.0  2073454.0  6824277.0  0.0   
people_fully_vaccinated              3021.0   893378.0  3595095.0  1.0   
daily_vaccinations_raw               4192.0    92034.0   289560.0  0.0   
daily_vaccinations                   7891.0    60968.0   213914.0  1.0   
total_vaccinations_per_hundred       4984.0       10.0       18.0  0.0   
people_vaccinated_per_hundred        4423.0        8.0       12.0  0.0   
people_fully_vaccinated_per_hundred  3021.0        4.0        8.0  0.0   
daily_vaccinations_per_million       7891.0     2699.0     4163.0  0.0   

                                         25%       50%        75%          max  
total_vaccinations                   40452.0  276962.0  1303350.0  130473853.0  
people_vaccinated                    36780.0  249555.0  1007708.0   85472166.0  
people_fully_vaccinated              19019.0  111716.0   488937.0   46365515.0  
daily_vaccinations_raw                2738.0   13838.0    56702.0    4575496.0  
daily_vaccinations                    1026.0    6161.0    27794.0    2541597.0  
total_vaccinations_per_hundred           1.0       4.0       11.0        162.0  
people_vaccinated_per_hundred            1.0       3.0        9.0         90.0  
people_fully_vaccinated_per_hundred      0.0       2.0        3.0         72.0  
daily_vaccinations_per_million         363.0    1305.0     3211.0      54264.0

In [21]:
#Basic statistics (2)
object_lst = ['country', 'iso_code', 'date', 'vaccines', 'source_name', 'source_website']
df[object_lst].astype('object').describe().transpose()

count unique  \
country         8079    151   
iso_code        8079    146   
date            8079    102   
vaccines        8079     26   
source_name     8079     92   
source_website  8079    147   

                                                              top  freq  
country                                          Northern Ireland   101  
iso_code                                                      MNG   175  
date                                          2021-03-10 00:00:00   133  
vaccines             Moderna, Oxford/AstraZeneca, Pfizer/BioNTech  1984  
source_name                                    Ministry of Health  2569  
source_website  https://coronavirus.data.gov.uk/details/health...   505

**KEY FINDINGS:**

Firstly, there is a relatively large discrepancy between daily vaccinations and daily vaccinations (raw). This needs to be investigated further in a later notebook.

Secondly, let's look at some key aspects of the numerical data:
1) The lowest number for daily vaccinations and people fully vaccinated is 1 (a little bizarre, but possibly due to the fact that missing values will qquite often represent 'ZERO', as in not startedt yet).
1) 130 million total vaccinations is the highest observed number in a country.
1) 85 million people vaccinated (partially or fully) observed number in a country.
1) The minimum the dataset started with was 1 person fully vaccinated and 1 daily vaccination.
1) The country with the highest level of people fully vaccinated has achieved 72%.
1) The country with the highest level of people vaccinated has achieved 90%.

Thirdly, let's look at some key aspects of the categorical data:
1) On 10th March the largest number of countries (133) supplied data.
1) Scotland has supplied the most data (101 days).
1) The vaccine combination Moderna, Biontech, AstraZeneca was so far most popular.
1) Most observations were provided from the UK Government website.

Next, let's check out which countries have produced the lowest and highest numbers (and when).

In [22]:
print(f'The entry with the lowest number of daily vaccinations is: \n \n')

#Return rows where daily vaccinations = 1, i.e. min
df[['date', 'country', 'daily_vaccinations']]\
    .loc[df['daily_vaccinations'] == df['daily_vaccinations'].min()]

The entry with the lowest number of daily vaccinations is: 
 



date  country  daily_vaccinations
807 2020-12-29  Belgium                 1.0

In [23]:
print(f'The entries with the lowest number of fully vaccinated people are: \n \n')

#Return rows where people fully vaccinated = 1, i.e. min
df[['date', 'country', 'people_fully_vaccinated']]\
    .loc[df['people_fully_vaccinated'] == df['people_fully_vaccinated'].min()]

The entries with the lowest number of fully vaccinated people are: 
 



date         country  people_fully_vaccinated
2458 2021-01-29  Faeroe Islands                      1.0
46   2021-02-02         Albania                      1.0
7601 2021-03-17         Ukraine                      1.0
7603 2021-03-19         Ukraine                      1.0
7604 2021-03-20         Ukraine                      1.0
7605 2021-03-21         Ukraine                      1.0
7606 2021-03-22         Ukraine                      1.0
7607 2021-03-23         Ukraine                      1.0

In [24]:
print(f'The entry with the highest number of total vaccinations is: \n \n')

#Return row with highest number for total vaccinations
df[['date', 'country', 'total_vaccinations']]\
    .loc[df['total_vaccinations'] == df['total_vaccinations'].max()]

The entry with the highest number of total vaccinations is: 
 



date        country  total_vaccinations
7882 2021-03-24  United States         130473853.0

In [25]:
print(f'The entry with the highest number of total vaccinations (%) is: \n \n')

#Return row with highest number for total vaccinations (%)
df[['date', 'country', 'total_vaccinations_per_hundred']]\
    .loc[df['total_vaccinations_per_hundred'] == df['total_vaccinations_per_hundred'].max()]

The entry with the highest number of total vaccinations (%) is: 
 



date    country  total_vaccinations_per_hundred
2876 2021-03-24  Gibraltar                          162.02

In [26]:
print(f'The entry with the highest number of daily vaccinations is: \n \n')

#Return row with highest number for total vaccinations
df[['date', 'country', 'daily_vaccinations']]\
    .loc[df['daily_vaccinations'] == df['daily_vaccinations'].max()]

The entry with the highest number of daily vaccinations is: 
 



date        country  daily_vaccinations
7871 2021-03-13  United States           2541597.0

In [27]:
print(f'The entries with the highest number of daily vaccinations (ppm) are: \n \n')

#Return row with highest number for total vaccinations
df[['date', 'country', 'daily_vaccinations_per_million']]\
    .loc[df['daily_vaccinations_per_million'] == df['daily_vaccinations_per_million'].max()]

The entries with the highest number of daily vaccinations (ppm) are: 
 



date           country  daily_vaccinations_per_million
2512 2021-02-08  Falkland Islands                         54264.0
2513 2021-02-09  Falkland Islands                         54264.0
2514 2021-02-10  Falkland Islands                         54264.0
2515 2021-02-11  Falkland Islands                         54264.0
2516 2021-02-12  Falkland Islands                         54264.0
2517 2021-02-13  Falkland Islands                         54264.0
2518 2021-02-14  Falkland Islands                         54264.0
2519 2021-02-15  Falkland Islands                         54264.0

In [28]:
print(f'The entry with the highest number of people fully vaccinated is: \n \n')

#Return row with highest number for people fully vaccinated
df[['date', 'country', 'people_fully_vaccinated']]\
    .loc[df['people_fully_vaccinated'] == df['people_fully_vaccinated'].max()]

The entry with the highest number of people fully vaccinated is: 
 



date        country  people_fully_vaccinated
7882 2021-03-24  United States               46365515.0

In [29]:
print(f'The entry with the highest number of people fully vaccinated (%) is: \n \n')

#Return row with highest number for people fully vaccinated (%)
df[['date', 'country', 'people_fully_vaccinated_per_hundred']]\
    .loc[df['people_fully_vaccinated_per_hundred'] == df['people_fully_vaccinated_per_hundred'].max()]

The entry with the highest number of people fully vaccinated (%) is: 
 



date    country  people_fully_vaccinated_per_hundred
2876 2021-03-24  Gibraltar                                72.29

In [30]:
print(f'The entry with the highest number of people vaccinated [full and partial] is: \n \n')

#Return row with highest number for people vaccinated
df[['date', 'country', 'people_vaccinated']]\
    .loc[df['people_vaccinated'] == df['people_vaccinated'].max()]

The entry with the highest number of people vaccinated [full and partial] is: 
 



date        country  people_vaccinated
7882 2021-03-24  United States         85472166.0

In [31]:
print(f'The entry with the highest number of people vaccinated (%) [full and partial] is: \n \n')

#Return row with highest number for people vaccinated (%)
df[['date', 'country', 'people_vaccinated_per_hundred']]\
    .loc[df['people_vaccinated_per_hundred'] == df['people_vaccinated_per_hundred'].max()]

The entry with the highest number of people vaccinated (%) [full and partial] is: 
 



date    country  people_vaccinated_per_hundred
2876 2021-03-24  Gibraltar                          89.73

**KEY FINDINGS:**

1) The United States of America vaccinate most in terms of absolute numbers.
1) Gibraltar vaccinates most in term of percentage of popuation.
1) The Falkland Islands have the highest number of daily vaccinations per million.

As a next step, let's take a look at how many, and which countries are included in the dataset.

In [32]:
#Number of countries included in dataset:
print(F'{df.country.nunique()} countries are included in the COVID-19 vaccination progress dataset.')

151 countries are included in the COVID-19 vaccination progress dataset.


In [33]:
#Alphabetically sorted list of unique country names
countries = df.country.unique()
print(f'The following countries are included in the dataset: \n \n {sorted(countries)}')

The following countries are included in the dataset: 
 
 ['Afghanistan', 'Albania', 'Algeria', 'Andorra', 'Angola', 'Anguilla', 'Antigua and Barbuda', 'Argentina', 'Australia', 'Austria', 'Azerbaijan', 'Bahamas', 'Bahrain', 'Bangladesh', 'Barbados', 'Belarus', 'Belgium', 'Belize', 'Bermuda', 'Bolivia', 'Brazil', 'Bulgaria', 'Cambodia', 'Canada', 'Cayman Islands', 'Chile', 'China', 'Colombia', 'Costa Rica', "Cote d'Ivoire", 'Croatia', 'Cyprus', 'Czechia', 'Denmark', 'Dominica', 'Dominican Republic', 'Ecuador', 'Egypt', 'El Salvador', 'England', 'Equatorial Guinea', 'Estonia', 'Faeroe Islands', 'Falkland Islands', 'Finland', 'France', 'Germany', 'Ghana', 'Gibraltar', 'Greece', 'Greenland', 'Grenada', 'Guatemala', 'Guernsey', 'Guinea', 'Guyana', 'Honduras', 'Hong Kong', 'Hungary', 'Iceland', 'India', 'Indonesia', 'Iran', 'Ireland', 'Isle of Man', 'Israel', 'Italy', 'Jamaica', 'Japan', 'Jersey', 'Jordan', 'Kazakhstan', 'Kenya', 'Kuwait', 'Laos', 'Latvia', 'Lebanon', 'Liechtenstein', 'Lithu

**KEY FINDINGS:**

1) 151 Countries are inlcuded in the dataset. The UN currently recognises 197 countries, of which 193 are members.
1) From visual inspection, the 'country' column appears to contain duplicates (England, Northern Ireland, Wales, Scotland are included in United Kingdom).

***
**SUMMARY OF IDENTIFIED CLEANING TASKS** 

- Drop 'daily_vaccinations_raw' column as only for cross-check purposes with daily vaccinations, i.e. information contained in total_vaccinations column

- Drop 'source_name' and 'source_website' as not required for data analysis (Note: Keep ISO for worldmap plots)

- Investigate and tidy-up 'vaccines' column

- Check 'country' column in detail as it contains duplicates (United Kingdom = England, Northern Ireland, Wales, Scotland)

NOTE: Columns with missing values will be checked in more detail  in a later notebook.

***

In [34]:
#Copy dataset before cleaning
df_1 = df.copy()

**Task 1 & 2:**

Drop 3 columns: 'daily_vaccinations_raw', 'source_name' and 'source_website'

In [35]:
#Drop columns daily_vaccinations_raw, source_name, source_website
lst = ['daily_vaccinations_raw', 'source_name', 'source_website']
df_1.drop(lst, 1, inplace=True)

print(f'The following {len(df_1.columns)} columns are included in the dataset: \n')
df_1.columns

The following 12 columns are included in the dataset: 



Index(['country', 'iso_code', 'date', 'total_vaccinations',
       'people_vaccinated', 'people_fully_vaccinated', 'daily_vaccinations',
       'total_vaccinations_per_hundred', 'people_vaccinated_per_hundred',
       'people_fully_vaccinated_per_hundred', 'daily_vaccinations_per_million',
       'vaccines'],
      dtype='object')

**TASK 3**:

Detailed investigation into vaccines column.

The information is stored as a list for each country (& date) containing all vaccines used up to the respective date. So, convert the vaccines column to a column per vaccine indicating whether a specific vaccine is used by a country (on a certain date).

In [36]:
#Take a look at how the vaccine a country uses are stored in the dataframe
print(f'The following {len(df_1.vaccines.unique())} vaccine groups are included in the original dataset: \n')
sorted(df_1.vaccines.unique())

The following 26 vaccine groups are included in the original dataset: 



['Covaxin, Oxford/AstraZeneca',
 'EpiVacCorona, Sputnik V',
 'Johnson&Johnson',
 'Johnson&Johnson, Moderna, Pfizer/BioNTech',
 'Moderna',
 'Moderna, Oxford/AstraZeneca',
 'Moderna, Oxford/AstraZeneca, Pfizer/BioNTech',
 'Moderna, Oxford/AstraZeneca, Pfizer/BioNTech, Sinopharm/Beijing, Sputnik V',
 'Moderna, Pfizer/BioNTech',
 'Oxford/AstraZeneca',
 'Oxford/AstraZeneca, Pfizer/BioNTech',
 'Oxford/AstraZeneca, Pfizer/BioNTech, Sinopharm/Beijing, Sinopharm/Wuhan, Sputnik V',
 'Oxford/AstraZeneca, Pfizer/BioNTech, Sinopharm/Beijing, Sputnik V',
 'Oxford/AstraZeneca, Pfizer/BioNTech, Sputnik V',
 'Oxford/AstraZeneca, Sinopharm/Beijing',
 'Oxford/AstraZeneca, Sinopharm/Beijing, Sputnik V',
 'Oxford/AstraZeneca, Sinovac',
 'Pfizer/BioNTech',
 'Pfizer/BioNTech, Sinopharm/Beijing',
 'Pfizer/BioNTech, Sinovac',
 'Pfizer/BioNTech, Sputnik V',
 'Sinopharm/Beijing',
 'Sinopharm/Beijing, Sinopharm/Wuhan, Sinovac',
 'Sinopharm/Beijing, Sputnik V',
 'Sinovac',
 'Sputnik V']

In [37]:
#Convert each entry in vaccines list to an individual row for a date/country
#and pivot to give dataframe no. 2 with individual vaccines, country and date columns
df_processed = df_1.drop('vaccines', axis=1).join(df['vaccines'].str.split(', ', expand=True).stack().reset_index(level=1, drop=True).rename('vaccines'))
df_processed = df_processed[['country', 'date', 'vaccines']].pivot(index=['country', 'date'], columns='vaccines', values='vaccines').reset_index()

#Create dataframe w/o country, date and vaccine columns
lst_ = ['country', 'date', 'vaccines']
df_2 = df_1.drop(lst_, 1)

#Join the two dataframes
df_processed = df_processed.join(df_2)

#Convert vaccine name entries to boolean values
#1stly create list of vaccines
v_lst = ', '.join(df.vaccines.tolist())
v_lst = list(set(v_lst.split(', ')))

#2ndly replace NaN values by 0
df_processed[v_lst] = df_processed[v_lst].fillna(0)

#Replace vaccine names by 1
vaccine_lst = ['Covaxin', 'EpiVacCorona', 'Johnson&Johnson', 'Moderna', 'Oxford/AstraZeneca', 'Pfizer/BioNTech',
                'Sinopharm/Beijing', 'Sinopharm/Wuhan', 'Sinovac', 'Sputnik V']
for c in vaccine_lst:
    df_processed[c] = df_processed[c].replace([c],1)
    
#Convert column header to pythonic standard (lowercase, no special characters)
df_processed.columns = df_processed.columns.str.lower().str.replace('&', '_').str.replace('/', '_').str.replace(' ', '_')

#Display sample from current dataframe
print(f'Sample of newly created vaccine columns: \n')
df_processed.iloc[:, :12].sample(1, random_state=42)

Sample of newly created vaccine columns: 



country       date  covaxin  epivaccorona  johnson_johnson  moderna  \
2809  Gibraltar 2021-01-16        0             0                0        0   

      oxford_astrazeneca  pfizer_biontech  sinopharm_beijing  sinopharm_wuhan  \
2809                   0                1                  0                0   

      sinovac  sputnik_v  
2809        0          0

**TASK 4**:

Detailed investigation of 'country' column and remove duplicates (England, Northern Ireland, Wales, Scotland).

In [38]:
#Spotcheck that the numbers for UK are equals to the sum for England, Scotland, Wales and northern Ireland.
lst2 = ['England', 'Scotland', 'Wales', 'Northern Ireland']
check1 = df.loc[df['country'].isin(lst2)].loc[df['date'] == '2021-03-15'].loc[:, 'daily_vaccinations'].sum()
check2 = df.loc[df['country'] == 'United Kingdom'].loc[df['date'] == '2021-03-15'].loc[:, 'daily_vaccinations'].values[0]
print(f'SPOTCHECK: The sum of daily vaccinations in England, Scotland, Wales and Nothern Ireland on 15th March 2021 was: {check1}')
print(f'SPOTCHECK: The total daily vaccinations in the United Kingdom on 15th March 2021 was: {check2}')

SPOTCHECK: The sum of daily vaccinations in England, Scotland, Wales and Nothern Ireland on 15th March 2021 was: 389942.0
SPOTCHECK: The total daily vaccinations in the United Kingdom on 15th March 2021 was: 389942.0


Since the spot-check showed that United Kingdom duplicates England, Scotland, Wales, Northern Ireland (as expected) rows containing the latter 4 were dropped (see code cell below) to give a total of 147 countries.

In [39]:
#Delete all rows for England, Scotland, Wales & Northern Ireland
df_processed = df_processed[df_processed.country.apply(lambda x: x not in ['England', 'Scotland', 'Wales', 'Northern Ireland'])]

As 147 countries are difficult to visualise, the countries will be grouped into regions (feature engineering section), keeping those we wish to investigate further separate, e.g. Asia, Europe, but UK & British Overseas, China, Japan, Israel, Russia, USA.

***

**SUMMARY OF UPCOMING FEATURE ENGINEERING TASKS** 

- Add new column for daily vaccinations in %

- Add new column for weekdays

- Add partially vaccinated column (total and %)

- Add vaccine type column

- Group countries, e.g. by continent

*NOTE: Additional features, such as population and GDP were aquired from an additional source and added in a different notebook.*

***

### 2.3 Feature Engineering to Prepare for In-Depth Data Analysis

In this section we undertake feature engineering to generate 6 new columns as outlined above.

**TASK 1**:

Create new column for daily vaccinations per hundred, i.e. percentage of population.

In [40]:
#Add new column for daily vaccinations in %
df_processed['daily_vaccinations_per_hundred'] = df_processed['daily_vaccinations_per_million']/10000

#Depict sample row
df_processed[['date', 'country', 'daily_vaccinations_per_hundred']].sample(1, random_state=42)

date     country  daily_vaccinations_per_hundred
6761 2021-03-21  Seychelles                          0.5206

**TASK 2**:

Create new column containing the day of the week, as the number of daily vaccinations might show a recurring pattern by weekday.

In [41]:
#Add new column for containing the weekday
df_processed['weekday'] = df_processed.date.dt.day_name()

#Depict sample row
df_processed[['date', 'country', 'weekday']].sample(1, random_state=42)

date     country weekday
6761 2021-03-21  Seychelles  Sunday

**TASK 3**:

Add new column reflecting 4 types of vaccines: inactivated virus, recombinant (viral vector), messenger RNA, protein subunit.

In [42]:
#Generate new columns for the four types of vaccines
df_processed['inactivated_virus'] = df_processed.covaxin | df_processed.sinopharm_beijing | df_processed.sinopharm_wuhan
df_processed['recombinant'] = df_processed.johnson_johnson | df_processed.oxford_astrazeneca | df_processed.sinovac | df_processed.sputnik_v
df_processed['rna'] = df_processed.pfizer_biontech | df_processed.moderna
df_processed['protein'] = df_processed.epivaccorona

#Depict sample row
df_processed[['inactivated_virus', 'recombinant', 'rna', 'protein']].sample(1, random_state=42)

inactivated_virus  recombinant  rna  protein
6761                  1            1    0        0

**TASK 4**:

Create 2 new columns reflecting partially vaccinated people (total and % values).

NOTE:
As a country's vaccination campaign will start with a first dose of vaccine, followed by a second dose with a delay of about 2-6 weeks, it can be hypothesised that there will be entries with data for people vaccinated, but missing data for fully vaccinated. Therefore, it can safely be assumed that a missing value (NaN) for 'fully vaccinated' translates to ZERO if there is data for 'people vaccinated' reported for that specific country on that date. This has been cross-checked for a selection of countries (Germany, United Kingdom, United States, Israel, India, Russia). An example (Germany) is depicted below and ence, 'partially vaccinated' will be calculated by inferring ZERO for NaN in case of missing values for fully vaccinated.

In [43]:
#Spot check examples when fully vaccinated status is not suplied for selected countries
df_check_fv = df_processed[['date', 'people_vaccinated', 'people_fully_vaccinated']][df_processed['country']=='Germany'] #Different countries can be entered
df_check_fv.set_index('date').loc['2021-01-12':'2021-01-16'] #Different date ranges can be checked

people_vaccinated  people_fully_vaccinated
date                                                  
2021-01-12           801778.0                      NaN
2021-01-13           901062.0                      NaN
2021-01-14          1000963.0                    114.0
2021-01-15          1093288.0                    545.0
2021-01-16          1150043.0                    945.0

In [44]:
#Generate new columns reflecting partially vaccinated people (total)
df_processed['people_partially_vaccinated'] = df_processed['people_vaccinated'] - df_processed['people_fully_vaccinated']
df_processed['people_partially_vaccinated'].fillna(df_processed['people_vaccinated'])

#Generate new columns reflecting partially vaccinated people (%)
df_processed['people_partially_vaccinated_per_hundred'] = df_processed['people_vaccinated_per_hundred'] - df_processed['people_fully_vaccinated_per_hundred']
df_processed['people_partially_vaccinated_per_hundred'].fillna(df_processed['people_vaccinated_per_hundred'])

#Depict sample row
print(f'Example row for new feature PARTIALLY VACCINATED: \n')
df_processed[['people_partially_vaccinated', 'people_vaccinated', 'people_fully_vaccinated',\
             'people_partially_vaccinated_per_hundred', 'people_vaccinated_per_hundred',\
              'people_fully_vaccinated_per_hundred']].sample(1, random_state=2)

Example row for new feature PARTIALLY VACCINATED: 



people_partially_vaccinated  people_vaccinated  people_fully_vaccinated  \
4483                      10872.0            16131.0                   5259.0   

      people_partially_vaccinated_per_hundred  people_vaccinated_per_hundred  \
4483                                     1.74                           2.58   

      people_fully_vaccinated_per_hundred  
4483                                 0.84

**TASK 5**:

Create a new column reflecting the grouped countries:

Map countries to custom regions, such as continents, but keeping countries of interest separate, e.g. Germany, UK & British Overseas, China, Japan, Israel, Russia, USA.

In [49]:
#Map countries to custom regions
#Noteworthy dictionary entries based on news coverage: Europe (EU plus more, minus UK), UK & British Overseas, China, India, Japan, Russia, Israel
#Create lists with countries per region
Africa = ('Algeria','Angola',"Cote d'Ivoire",'Egypt','Equatorial Guinea','Ghana','Guinea','Kenya','Malawi','Maldives','Mali','Mauritius','Morocco',
          'Mozambique','Namibia','Niger','Nigeria','Rwanda','Sao Tome and Principe', 'Senegal','Seychelles','Sierra Leone','Somalia','South Africa',
          'South Sudan','Sudan','Swaziland','Tanzania','Togo','Tunisia','Uganda','Zambia','Zimbabwe')

Asia = ('Afghanistan','Azerbaijan','Bangladesh','Bhutan','Brunei','Cambodia','East Timor','Hong Kong','Indonesia','Kazakhstan','Kyrgyzstan','Laos',
        'Macao','Malaysia','Maldives','Mongolia','Myanmar','Nepal','North Korea','Pakistan','Philippines','Singapore','South Korea','Sri Lanka',
        'Taiwan','Tajikistan','Thailand','Turkmenistan','Uzbekistan','Vietnam')
    
Middle_East = ('Bahrain','Iran','Iraq','Jordan','Kuwait','Lebanon','Oman','Qatar','Saudi Arabia','Syria','Turkey','United Arab Emirates','Yemen')
   
Europe = ('Albania','Andorra','Armenia','Austria','Belarus','Belgium','Bosnia and Herzegovina','Bulgaria','Croatia','Cyprus','Czechia','Denmark','Estonia',
          'Faeroe Islands','Finland','France','Georgia', 'Greece','Greenland','Hungary','Iceland','Ireland','Italy','Latvia','Liechtenstein',
          'Lithuania','Luxembourg','Macedonia','Malta','Moldova','Monaco','Montenegro','Netherlands','North Macedonia','Northern Cyprus','Norway','Poland',
          'Portugal','Romania','San Marino','Serbia','Slovakia','Slovenia','Spain','Sweden','Switzerland','Ukraine','Vatican City')

UK_BritishOversea= ('United Kingdom','Anguilla','Antigua and Barbuda','Bermuda','Cayman Islands','Falkland Islands','Gibraltar','Guernsey','Isle of Man',
                    'Jersey', 'Montserrat', 'Saint Helena', 'Turks and Caicos Islands')

North_America = ('Canada','United States')

Central_America_Caribic = ('Antigua and Barbuda','Bahamas','Barbados','Belize','Costa Rica','Cuba','Dominica','Dominican Republic','El Salvador','Grenada',
                           'Guatemala','Haiti','Honduras','Jamaica','Mexico','Nicaragua','Panama','Saint Kitts and Nevis','Saint Lucia',
                           'Saint Vincent and the Grenadines','Trinidad and Tobago')

South_America = ('Argentina','Bolivia','Brazil','Chile','Colombia','Ecuador','Guyana','Paraguay','Peru','Suriname','Uruguay','Venezuela')

Australia_Oceania = ('Australia','Fiji','Kiribati','Marshall Islands','Micronesia','Nauru','New Zealand','Palau','Papua New Guinea','Samoa','Solomon Islands',
                     'Tonga','Tuvalu','Vanuatu')

#Create dictionaries for mapping (incl. specific individual countries)
#NOTE: Future work - possibly keep United Arab Emirates separately as they are a key player in vaccintions
continents = {'China': 'China', 'Germany': 'Germany', 'India': 'India', 'Israel': 'Israel', 'Japan': 'Japan', 'Russia': 'Russia'} #individual countries
continents.update({country: 'Asia' for country in Asia})
continents.update({country: 'Europe' for country in Europe})
continents.update({country: 'Africa' for country in Africa})
continents.update({country: 'Middle_East' for country in Middle_East})
continents.update({country: 'UK_BritishOversea' for country in UK_BritishOversea})
continents.update({country: 'North_America' for country in North_America})
continents.update({country: 'Central_America_Caribic' for country in Central_America_Caribic})
continents.update({country: 'South_America' for country in South_America})
continents.update({country: 'Australia_Oceania' for country in Australia_Oceania})

#Create new column and map region to country:
df_processed['country_grouped'] = df_processed['country'].map(continents)

In [46]:
#List of unique grouped-countries
country_grouped = df_processed['country_grouped'].unique()
print(f'The following {len(country_grouped)} country-groups are now included in the dataset: \n \n {country_grouped}')

The following 15 country-groups are now included in the dataset: 
 
 ['Asia' 'Europe' 'Africa' 'UK_BritishOversea' 'Central_America_Caribic'
 'South_America' 'Australia_Oceania' 'Middle_East' 'North_America' 'China'
 'Germany' 'India' 'Israel' 'Japan' 'Russia']


### 2.4 Global Vaccination Dataframe

In this section we create a global vaccination dataframe for merging with the Twitter dataframe.

In [47]:
#Reduce dataset to appropriate columns
#NOTE: % columns ommited as global values cannot be calculated from this dataset!
#For colculations regarding % values the population per country is required (see different notebook
cols = ['date', 'covaxin', 'epivaccorona', 'johnson_johnson', 'moderna', 'oxford_astrazeneca',
        'pfizer_biontech', 'sinopharm_beijing', 'sinopharm_wuhan', 'sinovac', 'sputnik_v',
        'total_vaccinations', 'people_vaccinated', 'people_fully_vaccinated', 'daily_vaccinations',
        'people_partially_vaccinated']
df_global = df_processed[cols]

#Sum up all columns to get global vaccination data
df_global = df_global.groupby('date').sum()

***
***

## 3. SUMMARY

First, we will save the 2 dataframes we generated in order to be able to use them in other notebooks.

In [48]:
#Save dataframes for use in new notebooks
#Cleaned dataframe (set index to date and save)
df_processed.set_index('date').to_csv('data/df_processed.csv')

#Global dataframe
df_global.to_csv('data/df_global.csv')

1) Our PROCESSED dataframe contains the following:

| FEATURE | description | comment |
|----|----|----|
| **date** | data entry (13th December 2020 to 24th March 2021) | - |
| **weekday** | day of week (e.g. Monday) | NEW column added during feature engineering |
| **country** | 147 countries | 4 duplicates removed, UN recognises 197 countries (193 are members) |
| **country_grouped** | 17 countries/groups | NEW column added during feature engineering |
| **4 x vaccine type columns** | inactivated_virus, recombinant, rna, protein | NEW columns added during feature engineering |
| **10 x vaccine columns** | name/company | NEW columns added during cleaning to replace vaccines column |
| **daily_vaccinations** | no. of vaccination for that date/country | some days only daily vaccinations, some days only the (cummulative) total |
| **daily_vaccinations_per_million** | ratio (in ppm) between vaccination number and total population for the current date in the country | - |
| **daily_vaccinations_per_hundred** | ratio (in percent) between vaccination number and total population for the current date in the country | NEW column added during feature engineering |
| **total_vaccinations** | absolute number of immunisations (jabs) in the country | some days only daily vaccinations, some days only the (cummulative) total |
| **total_vaccinations_per_hundred** | ratio (in percent) between vaccination number and total population up to the date in the country | - |
| **people_vaccinated** | total no. of people with immunisation (partial and full) | - |
| **people_vaccinated_per_hundred** | ratio (in percent) between population immunized (partial and full) and total population | - |
| **people_fully_vaccinated** | total no. of people with full immunisation (typically 2) | - |
| **people_fully_vaccinated_per_hundred** | ratio (in percent) between population fully immunized and total population | - |
| **people_partially_vaccinated** | total no. of people with partial immunisation (i.e. 1 jab) | NEW column added during feature engineering |
| **people_partially_vaccinated_per_hundred** | ratio (in percent) between population partially immunized and total population | NEW column added during feature engineering |


2) Our GLOBAL dataframe contains a reduced set of the above features:

+ date
+ VACCINES: covaxin, epivaccorona, johnson_johnson, moderna, oxford_astrazeneca, pfizer_biontech, sinopharm_beijing, sinopharm_wuhan, sinovac, sputnik_v
+ NUMERIC COLUMNS: total_vaccinations, people_vaccinated, people_fully_vaccinated, daily_vaccinations, people_partially_vaccinated

***

### 3.1 Key Findings

The dataset cover a timeframe of 101 days (2020-12-13 to 2021-03-24) for 147 countries across the globe. Currently 10 different vaccines of 4 different subtypes (inactivated viruses, recombinant viral vectors, messengerRNA and protein subunit based vaccines) are approved and available for vaccinations across the globe. However, it should be noted that not every vaccine is authorised in every country. 

Apart from the daily vaccination column the dataset contains 40-60% empties in numerical feature columns. These large numbers of missing values will pose a challenge for the time series specific analysis and will hence be dealt with in the next notebook. Futhermore, there is a relatively large discrepancy between daily vaccinations and daily vaccinations (raw). This will also be investigated in the next notebook.

In addition, 19 new features were generated to further aid analysis:
+ Weekdays
+ 4 x vaccine type columns
+ 10 x vaccine columns
+ Grouped countries
+ Daily vaccinations in %
+ People partially vaccinated (absolute and %)

A first glance at the data has shown:

1) On 10th March the largest number of countries (133) supplied data.
1) Scotland has supplied the most data (101 days).
1) The vaccine combination Moderna, Biontech, AstraZeneca was so far most popular.
1) Most observations were provided from the UK Government website.
1) The United States of America vaccinate most in terms of absolute numbers.
    1) 130M total vaccinations on 24th March
    1) 25M vaccinations on 13th March
    1) 85M people vaccinated (partially and fully) on 24th March
    1) 46M people vaccinated (fully) on 24th March
1) Gibraltar vaccinates most in term of percentage of popuation.
    1) 89% people vaccinated (partially or fully) on 24th March
    1) 72% people vaccinated (fully) on 24th March
1) The Falkland Islands have the highest number of daily vaccinations per million.
    1) 54T vaccinations each day during 8-15th February

***

### 3.2 Key Limitations

The vaccination dataset has the following limitations:

+ The dataset is only split per country, however not per vaccine AND country. This limits the analyses and forecastings possible with respect to the different vaccines.

+ The dataset contains many features with 40-60% missing values. This will not negatively affect analysis, but it will hamper forecasting.

***